src/plumpy/process_states.py

```diff
@@ -195,10 +195,12 @@ class Running(State):
     _running: bool = False
     _run_handle = None
 
-    def __init__(self, process: 'Process', run_fn: Callable[..., Any], *args: Any, **kwargs: Any) -> None:
+    def __init__(
+        self, process: 'Process', run_fn: Callable[..., Union[Awaitable[Any], Any]], *args: Any, **kwargs: Any
+    ) -> None:
         super().__init__(process)
         assert run_fn is not None
-        self.run_fn = run_fn
+        self.run_fn = ensure_coroutine(run_fn)
```

Comment:
@chrisjsewell suggested to do `await_me_maybe` https://github.com/aiidateam/plumpy/pull/272/files#r1257558025 that would avoid this but I did not any technical reason do it. The blog post he links is only arguing for it cause of cleanness of the code.

Reply:
> I assumed wrapping a blocking function is like writing async to the function, it then is just executed like a blocking function when used with await. 

That's true, the blog post mentions the "maybe await" pattern is mostly for async framework that can support the downstream app can write block function. If the operation is block function, then it is run in block manner.

>  I think this run_fn will become the continue_fn when it is recover from the Waiting state, which means all such xx_fn should be coroutines along the way.

From aiida-core point of view, this never happened, since the `continue_fn` is never set in the `aiida-core` Waiting class. The `def run` is used to create the initial `Created` state and used to transfer the aiida Process
into its own `Waiting` state(s).